src/adventure/adventurewidget.cpp

```diff
@@ -0,0 +1,165 @@
+
+    m_journalTextEdit->setReadOnly(true);
+    m_journalTextEdit->setOverwriteMode(true);
+    m_journalTextEdit->setUndoRedoEnabled(false);
+    m_journalTextEdit->setDocumentTitle("Adventure Journal Text");
+    m_journalTextEdit->setTextInteractionFlags(Qt::TextSelectableByMouse);
+    m_journalTextEdit->setTabChangesFocus(false);
+
+    const auto &settings = getConfig().integratedClient;
+
+    QTextFrameFormat frameFormat = m_journalTextEdit->document()->rootFrame()->frameFormat();
+    frameFormat.setBackground(settings.backgroundColor);
+    m_journalTextEdit->document()->rootFrame()->setFrameFormat(frameFormat);
+
+    QTextCharFormat blockCharFormat = m_journalTextCursor->blockCharFormat();
+    blockCharFormat.setForeground(settings.foregroundColor);
+    auto font = new QFont();
+    font->fromString(settings.font); // needed fromString() to extract PointSize
+    blockCharFormat.setFont(*font);
+    m_journalTextCursor->setBlockCharFormat(blockCharFormat);
+
+    addJournalEntry(DEFAULT_MSG);
+
+    auto *layout = new QVBoxLayout(this);
+    layout->setAlignment(Qt::AlignTop);
+    layout->setContentsMargins(0, 0, 0, 0);
+    layout->setSpacing(0);
+    layout->addWidget(m_journalTextEdit);
+
+    connect(&m_adventureTracker,
+            &AdventureTracker::sig_accomplishedTask,
+            this,
+            &AdventureWidget::slot_onAccomplishedTask);
+
+    connect(&m_adventureTracker,
+            &AdventureTracker::sig_achievedSomething,
+            this,
+            &AdventureWidget::slot_onAchievedSomething);
+
+    connect(&m_adventureTracker, &AdventureTracker::sig_died, this, &AdventureWidget::slot_onDied);
+
+    connect(&m_adventureTracker,
+            &AdventureTracker::sig_gainedLevel,
+            this,
+            &AdventureWidget::slot_onGainedLevel);
+
+    connect(&m_adventureTracker,
+            &AdventureTracker::sig_killedMob,
+            this,
+            &AdventureWidget::slot_onKilledMob);
+
+    connect(&m_adventureTracker,
+            &AdventureTracker::sig_receivedHint,
+            this,
+            &AdventureWidget::slot_onReceivedHint);
+}
+
+void AdventureWidget::slot_onAccomplishedTask(double xpGained)
+{
+    // Only record accomplishedTask if it actually has associated xp to avoid
+    // spam, since sometimes it co-triggers with achievemnt and can be redundant.
+    if (xpGained > 0.0) {
+        auto msg = QString(ACCOMPLISH_MSG) + QString(XP_SUFFIX).arg(formatXPGained(xpGained));
+        addJournalEntry(msg);
+    }
+}
+
+void AdventureWidget::slot_onAchievedSomething(const QString &achievement, double xpGained)
+{
+    auto msg = QString(ACHIEVE_MSG).arg(achievement);
+
+    if (xpGained > 0.0)
+        msg = msg + QString(XP_SUFFIX).arg(formatXPGained(xpGained));
+
+    addJournalEntry(msg);
+}
+
+void AdventureWidget::slot_onDied(double xpLost)
+{
+    auto msg = QString(DIED_MSG);
+
+    if (xpLost < 0.0)
+        msg = msg + QString(XP_SUFFIX).arg(formatXPGained(xpLost));
+
```

Comment:
This logic can be simplified greatly. Let's exit early and only create `msg` once without mutating?

Reply:
Yeah I am actually tweaking things here to (optionally) include TP as well as XP